src/borg/testsuite/repository.py

```diff
@@ -716,339 +793,370 @@ def _subtly_corrupted_hints_setup(self):
+        rename_index(repository.path, "index.100")
+        check(repository, status=True)
+        assert list_indices(repository.path) == ["index.3"]
+        get_objects(repository, 4)
+        assert {1, 2, 3, 4, 5, 6} == list_objects(repository)
+
+
+def test_crash_before_compact(repository):
+    with repository:
+        repository.put(H(0), fchunk(b"data"))
+        repository.put(H(0), fchunk(b"data2"))
+        # Simulate a crash before compact
+        with patch.object(Repository, "compact_segments") as compact:
+            repository.commit(compact=True)
+            compact.assert_called_once_with(0.1)
+
+    with reopen(repository) as repository:
+        check(repository, repair=True)
+        assert pdchunk(repository.get(H(0))) == b"data2"
+
+
+def test_hints_persistence(repository):
+    with repository:
+        repository.put(H(0), fchunk(b"data"))
+        repository.delete(H(0))
+        repository.commit(compact=False)
+        shadow_index_expected = repository.shadow_index
+        compact_expected = repository.compact
+        segments_expected = repository.segments
+
+    # close and re-open the repository (create fresh Repository instance) to
+    # check whether hints were persisted to / reloaded from disk
+    with reopen(repository) as repository:
+        repository.put(H(42), fchunk(b"foobar"))  # this will call prepare_txn() and load the hints data
+        # check if hints persistence worked:
+        assert shadow_index_expected == repository.shadow_index
+        assert compact_expected == repository.compact
+        del repository.segments[2]  # ignore the segment created by put(H(42), ...)
+        assert segments_expected == repository.segments
+
+
+def test_hints_behaviour(repository):
+    with repository:
+        repository.put(H(0), fchunk(b"data"))
+        assert repository.shadow_index == {}
+        assert len(repository.compact) == 0
+        repository.delete(H(0))
+        repository.commit(compact=False)
+        # now there should be an entry for H(0) in shadow_index
+        assert H(0) in repository.shadow_index
+        assert len(repository.shadow_index[H(0)]) == 1
+        assert 0 in repository.compact  # segment 0 can be compacted
+        repository.put(H(42), fchunk(b"foobar"))  # see also do_compact()
+        repository.commit(compact=True, threshold=0.0)  # compact completely!
+        # nothing to compact anymore! no info left about stuff that does not exist any more:
+        assert H(0) not in repository.shadow_index
+        # segment 0 was compacted away, no info about it left:
+        assert 0 not in repository.compact
+        assert 0 not in repository.segments
+
+
+@pytest.fixture()
+def remote_repository(tmpdir):
+    repository_location = Location("ssh://__testsuite__" + os.path.join(str(tmpdir), "repository"))
+    with RemoteRepository(repository_location, exclusive=True, create=True) as remote_repository:
+        yield remote_repository
```

Comment:
guess this needs an update.

for the `repository` fixture, we have learned that we **must not** enter the CM within the fixture, but just return the CM and do the entering in the test function (using `with` there).

Reply:
Makes sense! I had left this as it was because we did not have any 'reopen' calls in the remote repo tests, so figured it would be fine. However, I now see what you mean, and will make this change 👍